hal/src/nRF52840/sleep_hal.cpp

```diff
@@ -49,59 +49,106 @@
+                NVIC_SetPriority(RTC2_IRQn, newPriority);
+            }
         } else if (source->type == HAL_WAKEUP_SOURCE_TYPE_BLE) {
-            priority->blePriority = NVIC_GetPriority(SD_EVT_IRQn);
-            NVIC_SetPriority(SD_EVT_IRQn, newPriority);
+            if (!blePriority.bumped) {
+                blePriority.priority = NVIC_GetPriority(SD_EVT_IRQn);
+                blePriority.bumped = true;
+            }
+            if (newPriority < NVIC_GetPriority(SD_EVT_IRQn)) {
+                NVIC_SetPriority(SD_EVT_IRQn, newPriority);
+            }
             NVIC_EnableIRQ(SD_EVT_IRQn);
         } else if (source->type == HAL_WAKEUP_SOURCE_TYPE_LPCOMP) {
-            priority->lpcompPriority = NVIC_GetPriority(COMP_LPCOMP_IRQn);
-            NVIC_SetPriority(COMP_LPCOMP_IRQn, newPriority);
+            if (!lpcompPriority.bumped) {
+                lpcompPriority.priority = NVIC_GetPriority(COMP_LPCOMP_IRQn);
+                lpcompPriority.bumped = true;
+            }
+            if (newPriority < NVIC_GetPriority(COMP_LPCOMP_IRQn)) {
+                NVIC_SetPriority(COMP_LPCOMP_IRQn, newPriority);
+            }
             NVIC_EnableIRQ(COMP_LPCOMP_IRQn);
         } else if (source->type == HAL_WAKEUP_SOURCE_TYPE_USART) {
-            priority->usart0Priority = NVIC_GetPriority(UARTE0_UART0_IRQn);
-            NVIC_SetPriority(UARTE0_UART0_IRQn, newPriority);
+            if (!usart0Priority.bumped) {
+                usart0Priority.priority = NVIC_GetPriority(UARTE0_UART0_IRQn);
+                usart0Priority.bumped = true;
+            }
+            if (newPriority < NVIC_GetPriority(UARTE0_UART0_IRQn)) {
+                NVIC_SetPriority(UARTE0_UART0_IRQn, newPriority);
+            }
             nrf_uarte_int_enable(NRF_UARTE0, NRF_UARTE_INT_RXDRDY_MASK);
         } else if (source->type == HAL_WAKEUP_SOURCE_TYPE_NETWORK) {
             auto network = reinterpret_cast<const hal_wakeup_source_network_t*>(source);
             if (!(network->flags & HAL_SLEEP_NETWORK_FLAG_INACTIVE_STANDBY)) {
-                priority->usart1Priority = NVIC_GetPriority(UARTE1_IRQn);
-                NVIC_SetPriority(UARTE1_IRQn, newPriority);
+                if (!usart1Priority.bumped) {
+                    usart1Priority.priority = NVIC_GetPriority(UARTE1_IRQn);
+                    usart1Priority.bumped = true;
+                }
+                if (newPriority < NVIC_GetPriority(UARTE1_IRQn)) {
+                    NVIC_SetPriority(UARTE1_IRQn, newPriority);
+                }
                 nrf_uarte_int_enable(NRF_UARTE1, NRF_UARTE_INT_RXDRDY_MASK);
             }
         }
         source = source->next;
     }
 }
 
-static void unbumpWakeupSourcesPriority(const hal_wakeup_source_base_t* wakeupSources, const WakeupSourcePriorityCache* priority) {
+static void unbumpWakeupSourcesPriority(const hal_wakeup_source_base_t* wakeupSources) {
     auto source = wakeupSources;
     while (source) {
```

Comment:
There is probably no need now to walk the list. We can just check the appropriate `bumped` field.

Reply:
I'd persist to traverse the list, as in long term for certain wakeup source it might not need to be bumped the priority if its normal priority is sufficient to wake up device. And for certain wakeup sources we are not just unbumping the priority, but also do something specific to the peripheral. If we depends on the bump flag, we will miss something.